contracts/chain-adapters/SP1_Adapter.sol

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import "./interfaces/AdapterInterface.sol";
+
+contract DataStore {
+    error NotHubPool();
+    mapping(address => bytes) public dataForTarget;
+    address public immutable hubPool;
+
+    modifier onlyHubPool() {
+        if (msg.sender != hubPool) {
+            revert NotHubPool();
+        }
+        _;
+    }
+
+    constructor(address _hubPool) {
+        hubPool = _hubPool;
+    }
+
+    function storeData(address target, bytes calldata data) external onlyHubPool {
+        dataForTarget[target] = data;
+    }
+}
+
+/**
+ * @notice Stores data that can be relayed to L2 SpokePool using SP1 + Helios light clients.
+ */
+contract SP1_Adapter is AdapterInterface {
+    DataStore public immutable DATA_STORE;
+
+    constructor(DataStore _store) {
+        DATA_STORE = _store;
+    }
+
+    /**
+     * @notice Saves root bundle data in a simple storage contract that can be proven and relayed to L2.
+     * @param target Contract on the destination that will receive the message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes calldata message) external payable override {
+        DATA_STORE.storeData(target, abi.encode(address(this), target, message));
```

Comment:
Something we should discuss: to keep the cost of sending roots low as we scale to many more chains, we may want to specifically store roots in a way that is target-agnostic (but we would need to differentiate this from admin messages) to avoid storing it for each chain.

One way this could work is that we check the function signature of the admin message. If it's relayRootBundle, we store it without the target, and then each call for other succinct chains will not need to store this value.

Thoughts?

Reply:
I thought about this but this only can work if we're ok overwriting the stored value every time. I'm not sure if this is how the Helios light clients work, and whether this forces the EOA publishing the messages on the L2 side to publish state in some order. 

For example, imagine we update the store to store value A on block 1 and then update it again to store value B on block 2. If we publish the state B on the L2 SpokePool before we publish state A, then is it possible to still relate A and B over to the SpokePool?

If not, then we would have to store data in the Store using an [array](https://github.com/across-protocol/contracts/pull/916#discussion_r1990499915) and keep pushing to this array. And then there aren't really benefits to storing data target-agnostic

Reply:
Related reasoning why storing data target-agnostic could be dangerous https://github.com/across-protocol/contracts/pull/916#discussion_r1990505774